Add PBKDF1 to the legacy provider

jon-oracle opened the pull request:
Added the PKCS#5 key derivation function to the default provider. 

This work precedes some other work to add library context arguments to the PKCS12 APIs.

EDIT: Moved to the legacy provider instead as this algorithm is not recommended for new applications.

paulidale commented on providers/implementations/kdfs/pbkdf1.c:
```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 1999-2021 The OpenSSL Project Authors. All Rights Reserved.
```
1999-2021 really?

slontis replied:
This is the date on the p5_crypt.c file where some of this code is extracted from.

paulidale reviewed (commented):
Do we really want a very old legacy algorithm added?

I'd like to see some tests added to evp_test too.

slontis reviewed (commented):
Looks fine to me..
Someone else not from Oracle will need to approve..
NOTE: This KDF will be used within PKCS5_PBE_keyivgen() where equivalent code currently resides.

slontis commented:
Small compile error is relevant

slontis commented:
> Do we really want a very old legacy algorithm added?

This is a KDF so it really belongs in the provider.

paulidale commented:
Perhaps it could start in the legacy provider :)

t8m commented:
I agree with Paul that this really belongs to legacy provider.

slontis commented on crypto/evp/p5_crpt.c:
```diff
@@ -91,7 +91,7 @@ int PKCS5_PBE_keyivgen(EVP_CIPHER_CTX *cctx, const char *pass, int passlen,
                                             (char *)mdname, 0);
     *p = OSSL_PARAM_construct_end();
     if (EVP_KDF_CTX_set_params(kctx, params) != 1
-            || EVP_KDF_derive(kctx, out, kl) != 1)
```
you may be able to pass the params into the derive here..

paulidale replied:
It's possible as of last night.

slontis commented:
Is this still WIP? changes name if it is not and maybe do a p i n g.

mattcaswell commented on crypto/evp/p5_crpt.c:
```diff
@@ -73,43 +78,33 @@ int PKCS5_PBE_keyivgen(EVP_CIPHER_CTX *cctx, const char *pass, int passlen,
-    memcpy(iv, md_tmp + (16 - ivl), ivl);
+
+    memcpy(key, out, PKCS5_PBES1_KEY_IV_LENGTH);
+    memcpy(iv, out + PKCS5_PBES1_KEY_IV_LENGTH, PKCS5_PBES1_KEY_IV_LENGTH);
```
Why are these temporary copies necessary? Can't you just directly pass `out` and `out + PKCS5_PBES1_KEY_IV_LENGTH` for the key and iv in `EVP_CipherInit_ex` call below?

mattcaswell commented on test/evp_kdf_test.c:
```diff
@@ -312,6 +312,59 @@ static int test_kdf_hkdf_empty_salt(void)
+    OPENSSL_free(params);
+    OSSL_PROVIDER_unload(prov);
+    return ret;
+}
```
Can we have a similar test somewhere for the public API function `PKCS5_PBE_keyivgen` which you've made significant changes to in this PR, but there doesn't seem to be a test for it anywhere.

mattcaswell reviewed (changes requested):
This needs a CHANGES entry to at least mention: the introduction of a new EVP_KDF algorithm, the fact that it is in the legacy provider, and that PKCS5_PBE_keyivgen now needs the legacy provider loaded in order to function.

jon-oracle commented:
> This needs a CHANGES entry to at least mention: the introduction of a new EVP_KDF algorithm, the fact that it is in the legacy provider, and that PKCS5_PBE_keyivgen now needs the legacy provider loaded in order to function.

Rebased and added a note to CHANGES.

t8m commented:
I think we want to squash the commits before merging as those are mostly work in progress commits during the review.

paulidale reviewed (changes requested):
I'd like to see at least a test in the `evp_test`.
I'm not sure why we're persisting with `evp_kdf_test`...

Otherwise, it looks good.

paulidale reviewed (approved):
Perfect, thanks for the extra bit.

slontis commented:
The error in legacy CI might be valid?

slontis commented:
Use
````
> perl configdata.pm --dump
````

slontis commented:
You will notice that the md2 related .txt files for test_evp have this line

````
Availablein = legacy
Digest = MD2
Input = "a"
Output = 32ec01ec4a6dac72c0ab96fb34c0b5d1
````

It needs to be the first line in a test for it to work..

jon-oracle commented:
> You will notice that the md2 related .txt files for test_evp have this line
> 
> ```
> Availablein = legacy
> Digest = MD2
> Input = "a"
> Output = 32ec01ec4a6dac72c0ab96fb34c0b5d1
> ```
> 
> It needs to be the first line in a test for it to work..

Yes that's already in the test file and as the PBKDF1 algorithm is in legacy it does appear to be loaded. The issue is inside the KDF where ossl_prov_digest_load_from_params() is called and fails to find MD2 despite it being built and available (the evp_test MD2 tests run OK).

Is there any way to list the loaded algorithms for a provider or print some details of the fetch?

paulidale commented:
`util/wrap.pl apps/openssl list -digest-algorithms -provider legacy` ?

t8m commented:
You could also try to run just the evp_test with verbose output to see if there are any errors dumped which could give more hints.

jon-oracle commented:
> You could also try to run just the evp_test with verbose output to see if there are any errors dumped which could give more hints.

The problem is that the library context supplied in the internal call to fetch the digest comes from the provider context and appears to be empty:

```
    OSSL_LIB_CTX *provctx = PROV_LIBCTX_OF(ctx->provctx);

    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, provctx))
        return 0;
```

@slontis has a provisional fix in https://github.com/openssl/openssl/pull/14507

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

jon-oracle commented on providers/implementations/kdfs/pbkdf1.c:
```diff
@@ -0,0 +1,242 @@
+{
+    const OSSL_PARAM *p;
+    KDF_PBKDF1 *ctx = vctx;
+    OSSL_LIB_CTX *provctx = PROV_LIBCTX_OF(ctx->provctx);
```
Used just after as an argument to ossl_prov_digest_load_from_params(). The name is a bit confusing though - it should be a 'libctx'. I've changed this.

mattcaswell commented:
Once #14991 is approved and merged this should magically start working.

t8m commented:
CI failures are relevant.

mattcaswell commented:
This should fix at least one of the problems:

```` diff
diff --git a/providers/common/build.info b/providers/common/build.info
index ba5df595fe..a14bf09037 100644
--- a/providers/common/build.info
+++ b/providers/common/build.info
@@ -4,5 +4,7 @@ SOURCE[../libcommon.a]=provider_err.c provider_ctx.c
 $FIPSCOMMON=provider_util.c capabilities.c bio_prov.c digest_to_nid.c\
             securitycheck.c provider_seeding.c
 SOURCE[../libdefault.a]=$FIPSCOMMON securitycheck_default.c
-SOURCE[../liblegacy.a]=provider_util.c
+IF[{- !$disabled{module} && !$disabled{shared} -}]
+  SOURCE[../liblegacy.a]=provider_util.c
+ENDIF
 SOURCE[../libfips.a]=$FIPSCOMMON securitycheck_fips.c
````

slontis commented:
Looks like you have a memory leak in your test.. (Use valgrind locally)

slontis commented:
test_kdf_pbkdf1 needs to be skipped for the no_legacy case.

mattcaswell commented:
The memleak seen in the test output is due to #15269.

The other failures are due to a bug in the child libctx code. PR coming up for that. Since the memleak only occurs in an error path which we only hit because of the underlying child libctx bug, fixing that bug is sufficient for this PR.

mattcaswell commented:
> The other failures are due to a bug in the child libctx code. PR coming up for that.

PR in #15270. With that PR applied to this branch I see the tests that are failing in this PR pass (at least locally).

mattcaswell commented:
#15270 has been merged. Please rebase again!

slontis commented:
Jon - when you do a force push like this - can you please add a comment also. e.g I had to rebase- there were no changes made.

slontis reviewed (commented):
Can you also update doc/man7/OSSL_PROVIDER-legacy.pod

jon-oracle commented:
> Jon - when you do a force push like this - can you please add a comment also. e.g I had to rebase- there were no changes made. (Or please review the last 2 commits).

No worries.

This was a rebase + make update to update the fips checksums.

slontis commented:
I am not 100% sure but I think the checksums should not be added anymore.. (There is a different process now - these files will only be updated when a Release in done).

slontis commented on doc/man7/OSSL_PROVIDER-legacy.pod:
```diff
@@ -78,6 +78,14 @@ Disabled by default. Use I<enable-rc5> config option to enable.
+
+=over 4
+
+=item PBKDF1, defined in L<https://datatracker.ietf.org/doc/html/rfc8018#section-5.1>
```
Just ````=item PBKDF1```` is probably sufficient here..

slontis replied:
Can you resolve things when you fix them please..

slontis commented on providers/fips.checksum:
```diff
@@ -1 +1 @@
-b3dca5cc989c42b9e46c0e0b1738ff17b51ce825f0b87ae13b8f609a0840978f  providers/fips-sources.checksums
+25ebfe80438755a6a997fd7b76a2d30725c7be0ae73b9378d0daf5e444453afa  providers/fips-sources.checksums
```
You may be able to remove this file?

paulidale commented:
Confirming that the checksums aren't required anymore.

jon-oracle commented:
Removed checksums and updated doc entry.

slontis commented on doc/man3/PKCS5_PBE_keyivgen.pod:
```diff
@@ -79,7 +83,7 @@ PKCS5_pbkdf2_set, PKCS5_pbkdf2_set_ex, EVP_PBE_scrypt, EVP_PBE_scrypt_ex
 =head2 Key Derivation
 
-PKCS5_PBE_keyivgen() takes a password I<pass> of
+PKCS5_PBE_keyivgen() and PKCS5_PBE_keyivgen_ex() take a password I<pass> of
```
Looks good Jon..
One last thing.. Should you mention here the PBKDF1 legacy change maybe?
PKCS#5 PBKDF1 key derivation has been moved from PKCS5_PBE_keyivgen() into
   the legacy crypto provider as an EVP_KDF

jon-oracle replied:
Done, thanks.

slontis commented on doc/man3/PKCS5_PBE_keyivgen.pod:
```diff
@@ -134,6 +134,9 @@ The *_keyivgen() functions are typically used in PKCS#12 to encrypt objects.
 These functions make no assumption regarding the given password.
 It will simply be treated as a byte sequence.
 
+From OpenSSL 3.0 the PBKDF1 algorithm used in PKCS5_PBE_keyivgen() and
```
Sorry maybe the history section?

slontis reviewed (commented):
LGTM.. (I cant be the first approver due to COI)

t-j-h commented on crypto/evp/p5_crpt.c:
```diff
@@ -49,16 +56,14 @@ int PKCS5_PBE_keyivgen(EVP_CIPHER_CTX *cctx, const char *pass, int passlen,
 
     ivl = EVP_CIPHER_iv_length(cipher);
-    if (ivl < 0 || ivl > 16) {
+    if (ivl != PKCS5_PBES1_KEY_IV_LENGTH) {
```
You have changed the values here from 16 to 8 and also wired in that key and iv length must be the same and must be 8 bytes which isn't what the code had before. Why?

jon-oracle replied:
The PBES1 algorithm defined in https://datatracker.ietf.org/doc/html/rfc2898#section-6.1 states:

>       3. Separate the derived key DK into an encryption key K consisting
>          of the first eight octets of DK and an initialization vector IV
>          consisting of the next eight octets:
> 
>                  K   = DK<0..7> ,
>                  IV  = DK<8..15> .

This is the constraint I used to determine the acceptable key/IV lengths.

t-j-h replied:
And you are applying this in a way that alters the logic - as these are properties of the input cipher.
You are now wiring them to a fixed value which is not what the code did previously. 

It matters little as to what the RFC states for things like this - we have a routine that can operate using any cipher and you are changing the logic of how it processes the input cipher significantly - and that should not be happening in this sort of code update.

slontis replied:
Tim are you aware of an actual use case?  Or is this a theoretical argument? I realize this is a public interface.. (albeit a fairly crappy one)

t-j-h replied:
It is a public function that takes and cipher and people do unusual things and wiring this to only work for iv==8 and keylength==8 seems rather inappropriate when the PR isn't about reducing the functionality to just what the RFC states.
We don't claim to limit it and 

If we wanted to do that then it should have been explicitly noted and raised and signed off by the OTC - not just mixed up in the middle of a patch that is meant to just be moving code around.

It really stood out as an input parameter validation change that had nothing to do with the patch.

Looking at the history some asserts got turned into these checks in 2019 a05bf83c7964bb3928b323fe356b9f70f105036d and the original asserts from 2002 are in 54a656ef081f72a740c550ebd8099b40b8b5cde0

There also appears to have been RC4 128-bit usage in the code originally (although not in the more recent code).

jon-oracle replied:
Ok, sure, I can see that this being a public function restricts how much we can clean up the implementation.

I've re-introduced the original logic now.

t-j-h replied:
Thanks.

slontis replied:
If I remember rightly I think we did this initially because it was trashing over bits of the buffer that it shouldnt if the size was > 8.. Oh well..

t-j-h replied:
That recollection would be in error. The code is not "trashing over bits of the buffer". If it did then that should be separately fixed. If we wanted to change the behaviour that should be separately fixed. When moving code around it is not the time to be altering behaviours like this as that makes for a challenge in terms of tracking changes.

The code was written in a manner that allowed for non 8-byte cipher algorithm use - there is evidence of that usage in the past - and that isn't covered in the RFC - but that isn't all that unusual with this RFCs which were written after-the-fact and are often simplifications or codifications of implementations that were more rigid that necessarily the practice out there in the wild.

The code has an explicit range check for both keylength and ivlength both of which are safe for the entire range of the check if you read the code. 

Although I see a separate potential issue - if the IV length is 16 bytes then we will use the key as the IV as well which is probably not exactly the expected behaviour. But there is no "trashing" that I can see. If you think there is then there should be a separate issue called out for this and it should be resolved independent of this PR.

slontis replied:
Yep. I didnt quite mean a buffer oveflow :)
md_tmp + (16 - ivl) is just wrong with these constraints..
The keylen constraint is the size of the buffer (64 bytes) - so there would be no space for iv in this case.

paulidale commented on test/recipes/05-test_pbe.t:
```diff
@@ -0,0 +1,27 @@
+$ENV{OPENSSL_MODULES} = abs_path(bldtop_dir("providers"));
+$ENV{OPENSSL_CONF} = abs_path(srctop_file("test", "default-and-legacy.cnf"));
+
+simple_test("test_pbe", "pbetest");
```
Better as: `ok(run(test(["pbetest"])), "running PBE test");` and not using Test::Simple and the $ENV{OPENSSL_MODULES} setting.

jon-oracle replied:
Does this include the legacy provider or is there another step needed?

paulidale reviewed (commented):
Is a test case with a fetched cipher useful?

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

paulidale commented on test/pbetest.c:
```diff
@@ -29,22 +32,30 @@ static unsigned char pbe_plaintext[] = {
+#endif
 
+#if !defined OPENSSL_NO_RC4 && !defined OPENSSL_NO_MD5
+#if  !defined OPENSSL_NO_DES && !defined OPENSSL_NO_SHA1
```
indentation space is in the wrong place

slontis commented on test/evp_kdf_test.c:
```diff
@@ -312,6 +312,61 @@ static int test_kdf_hkdf_empty_salt(void)
+    OSSL_PARAM *params = OPENSSL_malloc(sizeof(OSSL_PARAM) * 5);
+    OSSL_PARAM *p = params;
+
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD,
```
This doesnt do a NULL check if the malloc fails.

jon-oracle replied:
Fixed this and the other cases.

jon-oracle commented:
Rebased/squashed commits, removed changes to fips checksums. Added some review updates.

slontis commented on test/evp_kdf_test.c:
```diff
@@ -616,7 +616,7 @@ static int test_kdf_pbkdf2_small_iterations_pkcs5(void)
 {
     int ret = 0;
-    EVP_KDF_CTX *kctx;
+    EVP_KDF_CTX *kctx = NULL;
```
The same issue in in the Scrypt code below..

jon-oracle replied:
Scrypt test should be OK as it doesn't allocate params

slontis commented:
@paulidale - needs reapproval

openssl-machine commented:
This pull request is ready to merge

paulidale commented:
Oops, still need a second reviewer.

slontis commented:
Note libcrypto.num needs a rebase..

jon-oracle commented:
No worries, rebased.

paulidale commented:
And merged finally.  Thanks for the addition.
